modules/git/object_id.go

```diff
@@ -35,94 +31,54 @@ func (*Sha1Hash) Type() ObjectFormat { return Sha1ObjectFormat }
 
 var _ ObjectID = &Sha1Hash{}
 
-// EmptyObjectID creates a new ObjectID from an object format hash name
-func EmptyObjectID(objectFormatName string) (ObjectID, error) {
-	objectFormat := ObjectFormatFromName(objectFormatName)
-	if objectFormat != nil {
-		return objectFormat.EmptyObjectID(), nil
+func MustIDFromString(hexHash string) ObjectID {
+	id, err := NewIDFromString(hexHash)
+	if err != nil {
+		panic(err)
 	}
-
-	return nil, errors.New("unsupported hash type")
+	return id
 }
 
-func IDFromRaw(h ObjectFormat, b []byte) (ObjectID, error) {
-	if len(b) != h.FullLength()/2 {
-		return h.EmptyObjectID(), fmt.Errorf("length must be %d: %v", h.FullLength(), b)
+func NewIDFromString(hexHash string) (ObjectID, error) {
+	var theObjectFormat ObjectFormat
+	for _, objectFormat := range SupportedObjectFormats {
+		if len(hexHash) == objectFormat.FullLength() {
+			theObjectFormat = objectFormat
+			break
+		}
 	}
-	return h.MustID(b), nil
-}
 
-func MustIDFromString(h ObjectFormat, s string) ObjectID {
-	b, _ := hex.DecodeString(s)
-	return h.MustID(b)
-}
-
-func genericIDFromString(h ObjectFormat, s string) (ObjectID, error) {
-	s = strings.TrimSpace(s)
-	if len(s) != h.FullLength() {
-		return h.EmptyObjectID(), fmt.Errorf("length must be %d: %s", h.FullLength(), s)
+	if theObjectFormat == nil {
+		return nil, fmt.Errorf("length %d has no matched object format: %s", len(hexHash), hexHash)
 	}
-	b, err := hex.DecodeString(s)
+
+	b, err := hex.DecodeString(hexHash)
 	if err != nil {
-		return h.EmptyObjectID(), err
+		return nil, err
 	}
-	return h.NewID(b)
-}
 
-func IDFromString(hexHash string) (ObjectID, error) {
-	for _, objectFormat := range SupportedObjectFormats {
-		if len(hexHash) == objectFormat.FullLength() {
-			return objectFormat.NewIDFromString(hexHash)
-		}
+	if len(b) != theObjectFormat.FullLength()/2 {
```

Comment:
Must that assumption always be true?

Reply:
Yes, it MUST.

Reply:
Yes. Hex encoded format is 2 characters per encoded byte. Once it's decoded, it will be exactly half. The only way it would not be this is if you truncate the hash